plenum/recorder/src/replayer.py

```
@@ -0,0 +1,178 @@
+        Recorder(client_rec_kv_store, skip_metadata_write=True)
+
+
+def patch_sent_prepreapres(replaying_node, node_recorder):
+    sent_pps = {}
+    for k, v in node_recorder.store.iterator(include_value=True):
+        parsed = Recorder.get_parsed(v.decode())
+        outgoings = Recorder.filter_outgoing(parsed)
+        if outgoings:
+            for out in outgoings:
+                try:
+                    msg = json.loads(out[0])
+                    if isinstance(msg, dict) and OP_FIELD_NAME in msg:
+                        op_name = msg[OP_FIELD_NAME]
+                        if op_name == PREPREPARE and msg[f.INST_ID.nm] == 0:
+                            v, p = msg[f.VIEW_NO.nm], msg[f.PP_SEQ_NO.nm]
+                            sent_pps[v, p] = [msg[f.PP_TIME.nm],
+                                              [tuple(l) for l in
+                                               msg[f.REQ_IDR.nm]],
+                                              msg[f.DISCARDED.nm],
+                                              ]
+                        elif op_name == BATCH:
+                            for m in msg['messages']:
+                                try:
+                                    m = json.loads(m)
+                                    if m[OP_FIELD_NAME] == PREPREPARE and m[f.INST_ID.nm] == 0:
+                                        v, p = m[f.VIEW_NO.nm], m[
+                                            f.PP_SEQ_NO.nm]
+                                        sent_pps[v, p] = [m[f.PP_TIME.nm],
+                                                          [tuple(l) for l in
+                                                           m[f.REQ_IDR.nm]],
+                                                          m[f.DISCARDED.nm]
+                                                          ]
+                                except json.JSONDecodeError:
+                                    continue
```

We should not silently skip this error, as it means that recorder/replayer will not give us expected results.

---

This is to skip the hearbeat messages